### PR TITLE
infra: bump workspace to Rust edition 2024

### LINKED ISSUE #270: spec(infra): bump workspace to Rust edition 2024
## Overview

Onsager is on Rust edition 2021. The reasons are historical — when the
workspace was started, edition 2024 hadn't stabilized. Rust 1.95 (the
toolchain we pin) supports edition 2024 cleanly (stabilized in 1.85,
Feb 2025), and the Duhem foundation spec
([`onsager-ai/duhem#5`](https://github.com/onsager-ai/duhem/issues/5))
adopts edition 2024 from day one. Bringing Onsager onto the same edition
removes a small but persistent papercut: cross-repo reading of the same
person's code shouldn't require keeping two slightly-different edition
behaviors in your head.

The change itself is small. `[workspace.package].edition` is the single
source of truth (every member crate already inherits via
`edition.workspace = true`). `cargo fix --edition --workspace` handles
the idiom migration mechanically.

### What edition 2024 changes that matters here

- **Default prelude** now includes `Future` and `IntoFuture`. We can drop
  some `use std::future::Future;` lines `cargo fix --edition` won't catch
  — those are cleanup, not blockers.
- **`unsafe_op_in_unsafe_fn`** lints by default. Onsager has minimal
  `unsafe`; any new lints get explicit `unsafe { … }` blocks.
- **`if let` temporary scopes** narrow. Code that relied on the old
  scoping is rare; `cargo fix --edition` rewrites what it can.
- **Disjoint capture in closures** — already default since 2021; no-op.
- **Reserved keywords**: `gen`. We don't use that as an identifier today.

### Out of scope (not this spec)

- Toolchain pin bump (`rust-toolchain.toml` stays at `1.95.0`).
- Workspace dependency refresh.
- Clippy lint policy changes.
- Any subsystem-specific refactor surfaced by the migration — those
  earn their own spec if they're more than mechanical.

### Risk and rollback

Single-commit rollback. If `cargo build`/`test`/`clippy` go red and the
fix isn't trivially mechanical, we revert the workspace edition bump
and file a follow-up. The seam rule, file-budget lint, and event-
manifest lint are independent of edition and aren't affected.

## Plan

- [ ] Bump `[workspace.package].edition` from `"2021"` to `"2024"` in
      root `Cargo.toml`.
- [ ] Run `cargo fix --edition --workspace --allow-dirty` and stage the
      result.
- [ ] Reconcile any leftover warnings from edition 2024 defaults
      (`unsafe_op_in_unsafe_fn`, prelude shadowing). Mechanical fixes
      only — anything non-mechanical splits into a follow-up.
- [ ] Verify `cargo build --workspace`, `cargo test --workspace`,
      `cargo clippy --workspace --all-targets -- -D warnings`,
      `cargo fmt --all -- --check` all clean.
- [ ] Verify `xtask` lints (`check-file-budget`, `lint-seams`,
      `check-events`, `lint-api-contract`) still pass.

## Test

- [ ] `cargo build --workspace` clean on a fresh checkout.
- [ ] `cargo test --workspace` passes (or any failure traces to a
      pre-existing flake, recorded explicitly).
- [ ] `cargo clippy --workspace --all-targets -- -D warnings` clean.
- [ ] `cargo fmt --all -- --check` clean.
- [ ] `just lint` matches CI's verdict.
- [ ] CI workflow `rust` goes green on the PR.

### Human decides

- Already done — scope confirmed as edition-only (toolchain pin and
  dep refresh deferred).

### AI implements

- [ ] All Plan items.

### Open questions

> **Diffs from `cargo fix --edition` that are non-mechanical.** If the
> tool surfaces something it can't auto-rewrite (e.g. an `unsafe` fn
> body that needs explicit `unsafe { … }` blocks), the fix is mechanical
> and stays in this spec. If it surfaces something requiring real
> design judgement (rare for an edition bump), the change is split out
> and this spec ships only the mechanical part.

## Notes

- **Identity check.** Edition is a language-level toggle. Nothing in
  the four identity commitments at the top of `CLAUDE.md` is touched.
  Identity impact: no.
- **Why not bundle the toolchain bump?** Because it's a different
  signal. An edition bump is a deliberate language-version contract;
  a toolchain bump rolls clippy lint sets and may surface unrelated
  warnings. Keeping them in separate PRs makes the bisect surface
  cleaner if anything regresses.
- **Cross-repo coupling.** This decision was triggered by
  [`onsager-ai/duhem#5`](https://github.com/onsager-ai/duhem/issues/5)'s
  edition-2024 default. The Duhem spec lists Onsager-on-2021 as the
  *only* deliberate divergence; closing that gap unifies the toolchain
  story across both repos.

### PR DESCRIPTION
**Pull request by tikazyq**
## Summary

Bumps the Onsager workspace from Rust edition 2021 to edition 2024.
Triggered by [`onsager-ai/duhem#5`](https://github.com/onsager-ai/duhem/issues/5),
which adopts edition 2024 from day one for the new Duhem repo and
flagged Onsager-on-2021 as the only deliberate divergence — closing
that gap unifies the toolchain story across both repos.

The change itself is mostly mechanical:

- **`cargo fix --edition --workspace`** handled the language-level
  rewrites.
- **`cargo clippy --workspace --all-targets --fix`** collapsed 54
  nested `if let`s into edition-2024 `let`-chain form.
- **`cargo fmt --all`** applied the edition-2024 rustfmt style
  (import-ordering changes, where-clause indentation, etc.).

The only hand-written changes:

- **`crates/onsager-portal/src/dev_auth.rs` + `workflow_activation.rs`** —
  edition 2024 marks `std::env::set_var` / `remove_var` as `unsafe`
  (they aren't thread-safe). Six call sites in test code and the
  `ScopedEnvVar` helper get `unsafe { … }` blocks with a SAFETY note
  citing the surrounding `ENV_LOCK` discipline.
- **`crates/onsager-trigger/src/main.rs`** — match-ergonomics 2024
  disallows a redundant `ref mut` binding inside an
  implicitly-borrowed pattern. `cargo fix --edition` flagged it but
  couldn't auto-rewrite; the fix is dropping the modifier.
- **`CLAUDE.md`** — Conventions section bumped to "Rust edition 2024"
  so the document of record stays in sync.

Toolchain pin (`rust-toolchain.toml` 1.95.0) and workspace
dependencies are deliberately untouched — those carry independent
risk and would muddy the bisect surface if anything regresses.

## Test plan

Verified locally:

- [x] `cargo build --workspace` clean.
- [x] `cargo clippy --workspace --all-targets -- -D warnings` clean.
- [x] `cargo fmt --all -- --check` clean.
- [x] `cargo test --workspace --lib` — 114 passed, 0 failed.
- [x] `cargo test --workspace --bins` — 78 passed, 0 failed.
- [x] `cargo run -p xtask -- check-file-budget` clean.
- [x] `cargo run -p xtask -- lint-seams` clean.
- [x] `cargo run -p xtask -- check-events` clean.
- [x] `cargo run -p xtask -- check-triggers` clean.
- [x] `cargo run -p xtask -- check-api-contract` clean.
- [x] `cargo run -p xtask -- gen-event-docs --check` clean.
- [ ] CI `rust` workflow goes green (DB-backed integration tests in
      `tests/` need Postgres — left to CI).

Closes #270

---
_Generated by [Claude Code](https://claude.ai/code/session_01UTRotJUqw87KVnWyAekzeV)_

### CODE REVIEW

File: crates/onsager-portal/src/dev_auth.rs [hunk 1]
```diff
@@ -226,10 +226,13 @@ mod tests {
...
-        std::env::remove_var("USER");
+        // SAFETY: edition 2024 marks env mutation as `unsafe` because
+        // it isn't thread-safe. Single-threaded test, no other threads
+        // touch USER concurrently.
+        unsafe { std::env::remove_var("USER") };
         assert_eq!(dev_username(), "dev");
         if let Some(v) = prev {
-            std::env::set_var("USER", v);
+            unsafe { std::env::set_var("USER", v) };
         }
```
[the remark this answers is not in the document]
Comment (tikazyq) [reply crates/onsager-portal/src/dev_auth.rs]: Fixed in 4e9dd75 — refactored `dev_username` so the testable logic is a pure `dev_username_from(Option<String>)`, and the test now exercises that variant directly. No env mutation, no `unsafe`, no need for the shared lock. Also widened the test to cover the trim-empty fallback. Thanks — the SAFETY claim was indeed wrong (cargo's test harness is parallel by default).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/onsager-portal/src/workflow_activation.rs [hunk 2]
```diff
@@ -653,7 +653,10 @@ mod tests {
     impl ScopedEnvVar {
         fn set(key: &'static str, value: &str) -> Self {
             let previous = std::env::var(key).ok();
-            std::env::set_var(key, value);
+            // SAFETY: edition 2024 marks env mutation as `unsafe`. The
+            // surrounding tests serialize on `ENV_LOCK` so no other
+            // thread is reading or writing env concurrently.
+            unsafe { std::env::set_var(key, value) };
             Self { key, previous }
```
[the remark this answers is not in the document]
Comment (tikazyq) [reply crates/onsager-portal/src/workflow_activation.rs:661]: Fair point, but deferring to a follow-up. The `ENV_LOCK` + `ScopedEnvVar` pattern (lock acquired separately at the top of each test, then guards constructed below it) is pre-existing — edition 2024 only made the `unsafe` annotation explicit; it didn't introduce the discipline-by-convention shape. Centralizing into a shared `test_env` module where `ScopedEnvVar` holds the guard internally is a real improvement and I agree with the direction, but it's a refactor that goes beyond a "mechanical edition bump". Tracking as a follow-up; will spec separately if it doesn't already exist.

The other half of your concern — `dev_auth` mutating env without participating in this lock — is resolved by 4e9dd75, which removes env mutation from `dev_auth`'s test entirely.

---
_Generated by [Claude Code](https://claude.ai/code)_
